### ISSUE TITLE
Rust 1.48 support broken by `resolver` field in packaged `Cargo.toml`

### ISSUE THREAD

**Issue by davidhewitt** [serde-rs/serde#2255]
I have a [failure in PyO3 CI](https://github.com/PyO3/pyo3/runs/7740776066?check_suite_focus=true#step:8:33) which I can also reproduce locally. 

```
error: failed to download `serde v1.0.143`

Caused by:
  unable to get packages from source

Caused by:
  failed to parse manifest at `/Users/david/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.143/Cargo.toml`

Caused by:
  feature `resolver` is required
```

Looking at the unpacked `Cargo.toml` it contains `resolver = "1"` - I wonder if the `cargo package` process has somehow added this and broken support for older cargo versions?

**Comment by striezel** [comment]
Encountered the same issue in [one of my projects](https://github.com/striezel/corona/runs/7745203566?check_suite_focus=true).

Considering that serde 1.0.142 still worked with rustc 1.48 and seeing that serde 1.0.143 does not, let's take a look at the changes between those two versions: https://github.com/serde-rs/serde/compare/v1.0.142...v1.0.143
There is no change that adds `resolver = "1"` to the manifest in the code, so I guess this is probably an issue with packaging process. 🤔

**Comment by davidhewitt** [comment]
Looks like this is probably https://github.com/rust-lang/cargo/issues/10954

**Comment by davidhewitt** [comment]
Since [serde's own metadata](https://github.com/serde-rs/serde/blob/3d0251666e2af77091fff0d44f4b2eb127c9487b/serde/Cargo.toml#L15) claims to support 1.13+, I'll leave this open until the next serde release is pushed. Should help community members who hit this in the meanwhile.

**Comment by striezel** [comment]
> I followed up with @dtolnay on twitter DM about this, and his response was that there is no MSRV policy for serde, and thus this is an issue that is not likely to be fixed.

Hmm ... I find that a bit odd. :shrug: Sure, the readme could be argued about. But when even the cargo manifest shows something as rustc 1.13+, only to find out it actually requires something more like rustc 1.50+ when used in a procject feels a bit like false advertising to me. So either the readme _and_ manifest should be updated, or the crate should be rebuild with stable cargo which does not have that bug. There even is a dedicated CI job named `msrv` that tests with rustc 1.13. That does not feel like _"there is no MSRV policy for serde"_ to me. :confused:

I am aware that this may sound a bit harsh, especially considering that this issue is boils down to a bug in cargo and not in serde. So the serde developers are not even to blame for this. However, with such an important and widely used crate like serde this could potentially cause trouble for a lot of other projects that depend on serde. In my opinion, deciding not to fix this is not really helpful. So I am still hoping that one of the next serde releases comes with a fix for that - whether its just updating the MSRV or getting rid of that troublesome `resolver` line.

**Comment by Kijewski** [comment]
Where does the `resolver = "1"` line in the generated [Cargo.toml](https://docs.rs/crate/serde/1.0.143/source/Cargo.toml) come from, though? [Cargo.toml.orig](https://docs.rs/crate/serde/1.0.143/source/Cargo.toml.orig) does not contain it. Seems like a bug in `cargo package`, doesn't it?

**Comment by victorjulien** [comment]
In Suricata we try to support the Rust versions as they are shipped by the major (Linux) distributions, so Suricata can be packaged. For this reason we maintain a table of OS' that are important to us. Debian 11 (current stable) is the lowest, at Rust version 1.48. I think especially projects that are not pure Rust (like Suricata, which is C+Rust) care more about distro packaging and support, since we know our users do.

We're maintaining our table here: https://redmine.openinfosecfoundation.org/issues/4163 (screenshot added for reference)

![Screenshot from 2022-08-12 10-22-02](https://user-images.githubusercontent.com/32410611/184315625-7d48cc4a-0c4a-4fdc-b61e-078a8114630f.png)

I think would be great if Serde, and all other projects, would define a clear MSRV policy to help users of the crates ecosystem know what to expect.

**Comment by Kixunil** [comment]
Since there's a mismatch between `rust-version` in manifest and actual MSRV it could cause problems later if `cargo` becomes MSRV-aware and people start using it. It seems the best thing to do is to publish a new version with MSRV 1.13 and if `serde` MSRV is to be changed then do it in a single step involving update of the `rust-version` field. Also possibly yank 1.0.143.

**Comment by HamzaYasin1** [comment]
error: failed to run custom build command for `substrate-runtime v2.0.0 (/Users/hamza/Applications/Documents/blockchain/substrate-git/substrate/runtime)`
      
      Caused by:
        process didn't exit successfully: `/Users/hamza/Applications/Documents/blockchain/substrate-git/substrate/target/debug/build/substrate-runtime-29670f71a495afb9/build-script-build` (exit code: 1)
        --- stderr
            Updating crates.io index
        error: failed to download `serde v1.0.143`
      
        Caused by:
          unable to get packages from source
      
        Caused by:
          failed to parse manifest at `/Users/hamza/Applications/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.143/Cargo.toml`
      
        Caused by:
          feature `resolver` is required
      
          consider adding `cargo-features = ["resolver"]` to the manifest
    
    
    cargo 1.65.0-nightly (ce40690a5 2022-08-09)
    code base : substrate 2.0.0 

Having the same issue, i started getting from today, it was totally working fine couple of weeks back. I have used other versions of rust as well, but nothing was fruitful.
Command i'm using : cargo +nightly-2020-10-05 build

**Comment by taiki-e** [comment]
As said in #2259, the problem should be automatically resolved when the maintainer creates a new release using the nightly toolchain after nightly-2022-08-13 (tomorrow).

**Comment by taiki-e** [comment]
This has been fixed in 1.0.144: https://docs.rs/crate/serde/1.0.144/source/Cargo.toml

**Comment by striezel** [comment]
> This has been fixed in 1.0.144: https://docs.rs/crate/serde/1.0.144/source/Cargo.toml

Very good. :)

> I'm still _very_ confused as to the MSRV for serde. Should we be looking to rip serde out to get a consistent MSRV, or is the policy going to be made more clear?

I think the lesson here is that one can not really uphold a MSRV guarantee in one's own projects without pinning the dependencies down to exact versions via `Cargo.lock`, even if those dependencies have a certain MSRV themselves.
